pytest.skip() results in UnexpectedException error

In https://github.com/pytest-dev/pytest/pull/4927, it became possible to use `pytest.skip()` within a doctest.

<details><summary>Passing/skipping pytest output</summary>

```
$ pytest -sv --doctest-modules --pyargs niworkflows -k niworkflows.reports.core    
==================================================== test session starts =====================================================
platform linux -- Python 3.10.10, pytest-7.2.2, pluggy-1.0.0 -- /home/chris/mambaforge/envs/fmriprep-dev/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chris
plugins: doctestplus-0.12.1
collected 391 items / 387 deselected / 4 selected                                                                            

mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Report SKIPPED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Reportlet SKIPPED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Smallest PASSED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.run_reports SKIPPED

======================================== 1 passed, 3 skipped, 387 deselected in 3.32s ========================================
```

</details>

Adding `--doctest-plus` causes these skips to be reported as `UNEXPECTED EXCEPTION`:

<details><summary>Failing pytest call with `--doctest-plus`</summary>

```
$ pytest -sv --doctest-modules --pyargs niworkflows -k niworkflows.reports.core --doctest-plus
==================================================== test session starts =====================================================
platform linux -- Python 3.10.10, pytest-7.2.2, pluggy-1.0.0 -- /home/chris/mambaforge/envs/fmriprep-dev/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chris
plugins: doctestplus-0.12.1
collected 391 items / 387 deselected / 4 selected                                                                            

mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Report FAILED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Reportlet FAILED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Smallest PASSED
mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.run_reports FAILED

========================================================== FAILURES ==========================================================
_________________________________________ [doctest] niworkflows.reports.core.Report __________________________________________
248 
249 
250     .. testsetup::
251 
252         >>> cwd = os.getcwd()
253         >>> os.chdir(tmpdir)
254 
255         >>> from shutil import copytree
256         >>> from bids.layout import BIDSLayout
257         >>> test_data_path = find_resource_or_skip('data/tests/work')
UNEXPECTED EXCEPTION: Missing resource data/tests/work; run this test from a source repository
Traceback (most recent call last):
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest niworkflows.reports.core.Report[4]>", line 1, in <module>
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/conftest.py", line 45, in find_resource_or_skip
    pytest.skip(f"Missing resource {resource}; run this test from a source repository")
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/_pytest/outcomes.py", line 179, in skip
    raise Skipped(msg=reason, allow_module_level=allow_module_level)
Skipped: Missing resource data/tests/work; run this test from a source repository
/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py:257: UnexpectedException
________________________________________ [doctest] niworkflows.reports.core.Reportlet ________________________________________
099     object.
100 
101     .. testsetup::
102 
103         >>> cwd = os.getcwd()
104         >>> os.chdir(tmpdir)
105 
106         >>> from shutil import copytree
107         >>> from bids.layout import BIDSLayout
108         >>> test_data_path = find_resource_or_skip('data/tests/work')
UNEXPECTED EXCEPTION: Missing resource data/tests/work; run this test from a source repository
Traceback (most recent call last):
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest niworkflows.reports.core.Reportlet[4]>", line 1, in <module>
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/conftest.py", line 45, in find_resource_or_skip
    pytest.skip(f"Missing resource {resource}; run this test from a source repository")
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/_pytest/outcomes.py", line 179, in skip
    raise Skipped(msg=reason, allow_module_level=allow_module_level)
Skipped: Missing resource data/tests/work; run this test from a source repository
/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py:108: UnexpectedException
_______________________________________ [doctest] niworkflows.reports.core.run_reports _______________________________________
517 
518     Run the reports.
519 
520     >>> pytest.skip()
UNEXPECTED EXCEPTION: 
Traceback (most recent call last):
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest niworkflows.reports.core.run_reports[0]>", line 1, in <module>
  File "/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/_pytest/outcomes.py", line 179, in skip
    raise Skipped(msg=reason, allow_module_level=allow_module_level)
Skipped
/home/chris/mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py:520: UnexpectedException
================================================== short test summary info ===================================================
FAILED mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Report
FAILED mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.Reportlet
FAILED mambaforge/envs/fmriprep-dev/lib/python3.10/site-packages/niworkflows/reports/core.py::niworkflows.reports.core.run_reports
======================================== 3 failed, 1 passed, 387 deselected in 3.31s =========================================
```

</details>

My particular use-case is that I have some large test files in my source repository. I exclude them from the wheel/installed package to avoid excessively large `site-packages/` directories, so I've written fixtures/helper functions to call `pytest.skip()` if an excluded test file/directory is requested.

I had a quick look through your source code, and couldn't see a place to make a similar patch to the upstream one. Assuming this is something you'd be open to supporting, if you can point me in the right direction, I'm happy to submit a PR.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,8 @@
 0.12.2 (unreleased)
 ===================
 
+- Report doctests raising skip exceptions as skipped. [#196]
+
 0.12.1 (2022-09-26)
 ===================
 

--- a/pytest_doctestplus/plugin.py
+++ b/pytest_doctestplus/plugin.py
@@ -11,8 +11,10 @@ import sys
 import warnings
 from pathlib import Path
 from textwrap import indent
+from unittest import SkipTest
 
 import pytest
+from _pytest.outcomes import OutcomeException  # Private API, but been around since 3.7
 from packaging.version import Version
 
 from pytest_doctestplus.utils import ModuleChecker
@@ -258,7 +260,7 @@ def pytest_configure(config):
 
             # uses internal doctest module parsing mechanism
             finder = DocTestFinderPlus(doctest_ufunc=use_doctest_ufunc)
-            runner = doctest.DebugRunner(
+            runner = DebugRunnerPlus(
                 verbose=False, optionflags=options, checker=OutputChecker())
 
             for test in finder.find(module):
@@ -320,7 +322,7 @@ def pytest_configure(config):
 
             optionflags = get_optionflags(self) | FIX
 
-            runner = doctest.DebugRunner(
+            runner = DebugRunnerPlus(
                 verbose=False, optionflags=optionflags, checker=OutputChecker())
 
             parser = DocTestParserPlus()
@@ -707,3 +709,11 @@ class DocTestFinderPlus(doctest.DocTestFinder):
             tests = list(filter(test_filter, tests))
 
         return tests
+
+
+class DebugRunnerPlus(doctest.DebugRunner):
+    def report_unexpected_exception(self, out, test, example, exc_info):
+        cls, exception, traceback = exc_info
+        if isinstance(exception, (OutcomeException, SkipTest)):
+            raise exception
+        super().report_unexpected_exception(out, test, example, exc_info)

--- a/tests/test_doctestplus.py
+++ b/tests/test_doctestplus.py
@@ -965,6 +965,28 @@ def test_remote_data_ignore_warnings(testdir):
     testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(skipped=1)
 
 
+def test_skiptest(testdir):
+    testdir.makeini(
+        """
+        [pytest]
+        doctestplus = enabled
+    """
+    )
+    p = testdir.makepyfile(
+        """
+        class MyClass:
+            '''
+            >>> import pytest
+            >>> pytest.skip("I changed my mind")
+            >>> assert False, "This should not be reached"
+            '''
+            pass
+    """
+    )
+    reprec = testdir.inline_run(p, "--doctest-plus")
+    reprec.assertoutcome(skipped=1, failed=0)
+
+
 def test_ufunc(testdir):
     pytest.importorskip('numpy')
 

